### PR TITLE
Fix end-to-end timeout when a new node joins the network

### PR DESCRIPTION
**Pull request by jumaffre**
Resolves https://github.com/microsoft/CCF/issues/700

Our end-to-end test infra had a few limitations related to timeouts when a new node joins the network (notably with regards to https://github.com/microsoft/CCF/issues/703).

### CODE REVIEW

File: src/node/nodestate.h [hunk 1]
```diff
@@ -1049,25 +1049,8 @@ namespace ccf
...
-      {
-        LOG_FAIL_FMT("Could not open current service: status is not OPENING");
-        return false;
-      }
-
-      active_service->status = ServiceStatus::OPEN;
-      service_view->put(0, active_service.value());
-
-      return true;
+      GenesisGenerator g(network, tx);
```
Comment (jumaffre) [inline src/node/nodestate.h]: This is general refactoring that is unrelated to the goal of this PR.

File: tests/infra/clients.py [hunk 2]
```diff
@@ -261,13 +265,13 @@ def __init__(
         self.format = format
 
         while connection_timeout >= 0:
+            connection_timeout -= 0.1
             try:
                 self.connect()
                 break
-            except ssl.SSLError:
+            except (ssl.SSLError, ssl.SSLCertVerificationError):
```
Comment (jumaffre) [inline tests/infra/clients.py]: This exception was not caught here. It is raised when the peer's endorsing certificate does not match the one we expect (here `networkcert.pem`), related to changes implemented by https://github.com/microsoft/CCF/issues/552.

File: tests/infra/clients.py [hunk 3]
```diff
@@ -261,13 +265,13 @@ def __init__(
         self.format = format
 
         while connection_timeout >= 0:
+            connection_timeout -= 0.1
```
Comment (jumaffre) [inline tests/infra/clients.py]: This loop did not work properly. When `connection_timeout` was about to be decremented for the last time (i.e. `connection_timeout=0`) and `self.connect()` was failing, the exception would be caught, `connection_timeout` was still positive and we wouldn't `raise`. We would then decrement `connection_timeout` (now equals to `-0.1`) and the `while` loop would stop. We would then try to send a request, which would timeout "forever" (`ctest` actually times out after 240 seconds).

File: tests/infra/ccf.py [hunk 4]
```diff
@@ -21,6 +21,10 @@
 
 logging.getLogger("paramiko").setLevel(logging.WARNING)
 
+# TODO: Quote verification may take a long time to execute on the leader when joining
+# https://github.com/microsoft/CCF/issues/703
+JOIN_TIMEOUT = 30
```
Comment (achamayou) [inline tests/infra/ccf.py]: Should we make sure this is < to the raft election timeout, at least until we have multi-threaded execution?

File: tests/infra/clients.py (hunk 3 above)
Comment (achamayou) [reply tests/infra/clients.py]: Ah, good catch!

**Comment by cimetrics** [comment]
fix_e2e_timeout@[3871 aka 20200120.5](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3871&view=results) vs master ewma over 20 builds from [3742](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3742&view=results) to [3864](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3864&view=results)
images

**Comment by cimetrics** [comment]
fix_e2e_timeout@[3882 aka 20200120.11](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3882&view=results) vs master ewma over 20 builds from [3743](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3743&view=results) to [3880](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3880&view=results)
images

**Comment by cimetrics** [comment]
fix_e2e_timeout@[3888 aka 20200120.13](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3888&view=results) vs master ewma over 20 builds from [3743](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3743&view=results) to [3880](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3880&view=results)
images

File: tests/infra/ccf.py (hunk 4 above)
Comment (jumaffre) [reply tests/infra/ccf.py]: As discussed, the two timers are somewhat orthogonal for our end-to-end tests. The `JOIN_TIMEOUT` just needs to greater than the time it takes for the quote verification to execute (~25 secs for worst case).

**Comment by cimetrics** [comment]
fix_e2e_timeout@[3896 aka 20200120.17](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3896&view=results) vs master ewma over 20 builds from [3743](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3743&view=results) to [3880](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3880&view=results)
images

**Comment by cimetrics** [comment]
fix_e2e_timeout@[3898 aka 20200120.18](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3898&view=results) vs master ewma over 20 builds from [3743](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3743&view=results) to [3880](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3880&view=results)
images
